lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2645,6 +2645,46 @@ class OpaqueUnderlyingTypeChecker : public ASTWalker {
     // If there are no candidates, then the body has no return statements, and
     // we have nothing to infer the underlying type from.
     if (Candidates.empty()) {
+      // We try to find if the last element of the `Body` multi element
+      // `BraceStmt` is an expression that produces a value that satisfies all
+      // the opaque type requirements and if that is the case, it means we can
+      // suggest a fix-it note to add an explicit `return`.
+      if (Body->getNumElements() > 1) {
+        auto element = Body->getLastElement();
+        // Let's see if the last statement would make for a valid return value.
+        if (auto expr = element.dyn_cast<Expr *>()) {
+          bool conforms = true;
+
+          for (auto requirement :
+               OpaqueDecl->getOpaqueInterfaceGenericSignature().getRequirements()) {
+            auto requirementFulfilled =
+                TypeChecker::typesSatisfyConstraint(expr->getType()->getRValueType(),
```

Comment:
Ah, there is a `TypeChecker::conformsToProtocol` maybe could be simpler to use here?

Reply:
It saved one line, must have overlooked this one. Thanks again!